osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java

```diff
@@ -761,8 +768,84 @@ public ResponseMessage dequeueSwitchFirmwareResponse(final String correlationUid
         return this.commonResponseMessageFinder.findMessage(correlationUid);
     }
 
+    /**
+     * @param organisationIdentification
+     *            the organisation the device we want to check belongs to
+     * @param deviceId
+     *            the id of the device we are checking
+     * @param firmwareVersions
+     *            the list of firmware versions to check if they are in the
+     *            history of the devices firmware history
+     * @return a list of firmware versions not present in the the devices
+     *         firmware history
+     * @throws FunctionalException
+     */
+    public List<FirmwareVersionDto> checkFirmwareHistoryForVersion(final String organisationIdentification,
+            final String deviceId, final List<FirmwareVersionDto> firmwareVersions) throws FunctionalException {
+
+        // copy input parameter
+        final List<FirmwareVersionDto> firmwareVersionsToCheck = new ArrayList<>();
+        firmwareVersionsToCheck.addAll(firmwareVersions);
+
+        // get history
+        final List<DeviceFirmwareFile> deviceFirmwareFiles = this.getDeviceFirmwareFiles(organisationIdentification,
+                deviceId);
+        final List<FirmwareVersionDto> firmwareVersionsInHistory = deviceFirmwareFiles.stream()
+                .map(d -> d.getFirmwareFile().getModuleVersions().entrySet())
+                .flatMap(Collection::stream)
+                .map(e -> new FirmwareVersionDto(FirmwareModuleType.getByDescription(e.getKey().getDescription()),
+                        e.getValue()))
+                .collect(Collectors.toList());
+
+        // remove the history versions
+        firmwareVersionsToCheck.removeAll(firmwareVersionsInHistory);
+
+        return firmwareVersionsToCheck;
+    }
+
+    @Transactional(value = "writableTransactionManager")
+    public void tryToAddFirmwareVersionToHistory(final String organisationIdentification,
+            final String deviceIdentification, final FirmwareVersionDto firmwareVersion) throws FunctionalException {
+
+        final FirmwareModule module = createFirmwareModule(firmwareVersion);
+        final Device device = this.domainHelperService.findActiveDevice(deviceIdentification);
+        final List<FirmwareFile> firmwareFiles = this.getPossibleFirmwareFiles(organisationIdentification, device);
+
+        // check each file for the module and the version as returned by the
+        // device
+        for (final FirmwareFile file : firmwareFiles) {
+            final Map<FirmwareModule, String> moduleVersions = file.getModuleVersions();
+            if (moduleVersions.containsKey(module) && moduleVersions.get(module).equals(firmwareVersion.getVersion())) {
```

Comment:
For SSLDs the firmware files would only contain a single module, for smart meters I believe this might not always be the case. Would it then be possible to have multiple hits for a specific module version. and if so, should we take this into account here?

Reply:
only one record is stored now